runtime/parachains/src/reward_points.rs

```diff
@@ -23,33 +23,107 @@
 
 use primitives::v1::ValidatorIndex;
 use pallet_staking::SessionInterface;
+use crate::shared;
 
 /// The amount of era points given by backing a candidate that is included.
 pub const BACKING_POINTS: u32 = 20;
 
 /// Rewards validators for participating in parachains with era points in pallet-staking.
 pub struct RewardValidatorsWithEraPoints<C>(sp_std::marker::PhantomData<C>);
 
-fn reward_by_indices<C, I>(points: u32, indices: I) where
-	C: pallet_staking::Config,
+fn validators_to_reward<C, T, I>(validators: &'_ [T], indirect_indices: I) -> impl IntoIterator<Item=&'_ T> where
+	C: shared::Config,
 	I: IntoIterator<Item = ValidatorIndex>
 {
-	// Fetch the validators from the _session_ because sessions are offset from eras
-	// and we are rewarding for behavior in current session.
-	let validators = C::SessionInterface::validators();
-	let rewards = indices.into_iter()
-		.filter_map(|i| validators.get(i.0 as usize).map(|v| v.clone()))
-		.map(|v| (v, points));
-
-	<pallet_staking::Module<C>>::reward_by_ids(rewards);
+	let validator_indirection = <shared::Module<C>>::active_validator_indices();
+
+	indirect_indices.into_iter()
+		.filter_map(move |i| validator_indirection.get(i.0 as usize).map(|v| v.clone()))
+		.filter_map(move |i| validators.get(i.0 as usize))
```

Comment:
Could you clarify in which cases `filter_map` can return `None`? Is it considered an error?
My understanding is that the first `filter_map` can return `None` if `max_validators.is_some()` and the second should always be `Some`.

Reply:
Yes, that should be true, but there may be some inconsistency in length if storage is tampered with. In such instances we should not panic as that would interfere with the chain being saved by governance.

Reply:
I was not suggesting a panic, but logging an error if propagating it is not possible, which is slightly better than silently ignoring it.